File: backend/routes/generate.py
import json
import anthropic
from flask import Blueprint, request, Response, stream_with_context
from prompts import get_system_prompt, get_user_prompt

generate_bp = Blueprint('generate', __name__)

client = None


def get_client():
    global client
    if client is None:
        import os
        client = anthropic.Anthropic(api_key=os.environ['ANTHROPIC_API_KEY'])
    return client


def event(data: dict) -> str:
    return f"data: {json.dumps(data)}\n\n"


@generate_bp.post('/generate')
def generate():
    body = request.get_json(silent=True) or {}
    prompt = body.get('prompt', '').strip()
    template = body.get('template', '').strip()

    if not prompt or not template:
        return {'error': 'prompt and template are required'}, 400

    def stream():
        try:
            with get_client().messages.stream(
                model='claude-haiku-4-5',
                max_tokens=4096,
                system=[
                    {
                        'type': 'text',
                        'text': get_system_prompt(),
                        'cache_control': {'type': 'ephemeral'},  # 90% cheaper on repeat calls
                    }
                ],
                messages=[
                    {'role': 'user', 'content': get_user_prompt(template, prompt)}
                ],
            ) as s:
                for text in s.text_stream:
                    yield event({'text': text})

            yield event({'done': True})

        except Exception as e:
            print(f'Claude API error: {e}')
            yield event({'error': str(e)})

    return Response(
        stream_with_context(stream()),
        content_type='text/event-stream',
        headers={
            'Cache-Control': 'no-cache',
            'Connection': 'keep-alive',
            'X-Accel-Buffering': 'no',
        },
    )

File: backend/routes/generate.js
import express from 'express'
import Anthropic from '@anthropic-ai/sdk'
import { getSystemPrompt, getUserPrompt } from '../prompts/index.js'

export const generateRoute = express.Router()

let client
function getClient() {
  if (!client) client = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY })
  return client
}

generateRoute.post('/generate', async (req, res) => {
  const { prompt, template } = req.body

  if (!prompt || !template) {
    return res.status(400).json({ error: 'prompt and template are required' })
  }

  // Set up SSE streaming
  res.setHeader('Content-Type', 'text/event-stream')
  res.setHeader('Cache-Control', 'no-cache')
  res.setHeader('Connection', 'keep-alive')

  try {
    const stream = getClient().messages.stream({
      model: 'claude-haiku-4-5',          // cheapest model
      max_tokens: 4096,                    // skeleton fills need ~2000-3000 tokens output
      system: [
        {
          type: 'text',
          text: getSystemPrompt(),
          cache_control: { type: 'ephemeral' } // cache system prompt — 90% cheaper on repeat calls
        }
      ],
      messages: [
        {
          role: 'user',
          content: getUserPrompt(template, prompt)
        }
      ]
    })

    for await (const event of stream) {
      if (
        event.type === 'content_block_delta' &&
        event.delta.type === 'text_delta'
      ) {
        // Send each chunk to frontend as SSE
        res.write(`data: ${JSON.stringify({ text: event.delta.text })}\n\n`)
      }
    }

    res.write(`data: ${JSON.stringify({ done: true })}\n\n`)
    res.end()

  } catch (err) {
    console.error('Claude API error:', err.message)
    res.write(`data: ${JSON.stringify({ error: err.message })}\n\n`)
    res.end()
  }
})
